content/blog/2022-04-19-tracing/index.md

```diff
@@ -0,0 +1,142 @@
+
+A graph depicting this approach is shown below:
+
+<p align="center">
+<img src="nested_tree.png" alt="nested Tree Formation" title="nested Tree Formation" style="zoom:25%;" />
+</p>
+
+# TraceMonkey VM Implementation & Optimization
+
+Finally, we take a look at how TraceMonkey's VM implemented. TraceMonkey's VM stores all compiled traces inside trace cache, indexed by interpreter PC and type signatures. The trace monitor builds a activation record when executing a trace, with associated local and global variables, temporary stack spaces, and space for arguments to native calls. The trace could then be called like a normal function pointer. The trace monitor is also responsible for restoring the interpreter state, verifying states for compiled trace and releasing the occupied space.
+
+The author also introduces an intresting optimization technique called Tree Stiching. The purpose of tree stitching is to reuse activation record when a side exit is met. The reason is fairly straightforward, since it's easy to see that within a loop, the register allocation and memory layouts are often invariant. the exiting trace only needs to write live
+register-carried values back to its trace activation record. However, with multi-code processors being the majority (unlike when this paper just got published), it's probably just easier to recompile the entire trace-tree in backend.
+
+Additional optimization techniques when a LIR trace gets compiled to machine code as a set of filters, including
+1. Float to Integer Conversion
+2. Constant Subexpression Elimination
+3. Constant Folding
+4. Dead Code Elimination
+5. Dead Stack Elimination
+
+# Evaluation
+The evaluation of is paper is performed on a Macbook Pro over SunSpider, which is a standard testing suite for JavaScript with a variety of applications. The experimental environment is very reasonable to me, since it's a very good representation of JavaScript workloads and user's device. The author benchmarks TraceMonkey against SpiderMonkey (baseline interpreter), SFX (inline threading interpreter in Apple's Webkit), V8 (Method-based JIT developed by Google).
```

Comment:
> it's a very good representation of JavaScript workloads

Can you explain a little more about why you think it's representative of real JavaScript workloads? I actually think SunSpider might be more contrived than you're imagining. [Here's the actual code](https://github.com/WebKit/WebKit/blob/main/Websites/webkit.org/perf/sunspider-1.0.2/sunspider-1.0.2/sunspider-test-contents.js), if you're interested!